libtenzir/builtins/connectors/s3.cpp

```diff
@@ -21,28 +21,49 @@ namespace tenzir::plugins::s3 {
 
 namespace {
 
+struct s3_config {
+  std::string access_key = {};
+  std::string secret_key = {};
+  std::string session_token = {};
+
+  friend auto inspect(auto& f, s3_config& x) -> bool {
+    return f.object(x)
+      .pretty_name("s3_config")
+      .fields(f.field("access-key", x.access_key),
+              f.field("secret-key", x.secret_key),
+              f.field("session-token", x.session_token));
+  }
+};
+
 struct s3_args {
-  bool anonymous;
-  located<std::string> uri;
+  bool anonymous = {};
+  located<std::string> uri = {};
+  std::optional<s3_config> config = {};
 
   template <class Inspector>
   friend auto inspect(Inspector& f, s3_args& x) -> bool {
     return f.object(x).pretty_name("s3_args").fields(
-      f.field("anonymous", x.anonymous), f.field("uri", x.uri));
+      f.field("anonymous", x.anonymous), f.field("uri", x.uri),
+      f.field("config", x.config));
   }
 };
 
 auto get_options(const s3_args& args) -> arrow::fs::S3Options {
-  auto opts = arrow::fs::S3Options::Anonymous();
-  if (not args.anonymous) {
-    auto opts_from_uri = arrow::fs::S3Options::FromUri(args.uri.inner);
-    if (not opts_from_uri.ok()) {
-      opts = arrow::fs::S3Options::Defaults();
-    } else {
-      opts = *opts_from_uri;
-    }
+  auto opts = arrow::fs::S3Options::FromUri(args.uri.inner);
+  if (not opts.ok()) {
+    diagnostic::error("failed to parse S3 options: {}",
+                      opts.status().ToString())
+      .throw_();
+  }
```

Comment:
I changed this around so that it now shows the proper error message to the user. I also wrote a follow-up story to render thrown diagnostics in exec nodes properly.